File: day01/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"slices"
	"strconv"
	"strings"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	var nums1, nums2 []int

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		f := strings.Fields(scanner.Text())
		num1, err := strconv.Atoi(f[0])
		if err != nil {
			log.Fatalln(err)
		}
		num2, err := strconv.Atoi(f[1])
		if err != nil {
			log.Fatalln(err)
		}
		nums1 = append(nums1, num1)
		nums2 = append(nums2, num2)
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	slices.Sort(nums1)
	slices.Sort(nums2)

	minNum2 := nums2[0]
	maxNum2 := nums2[len(nums2)-1]
	counts := make([]int, maxNum2-minNum2+1)
	for _, i := range nums2 {
		counts[i-minNum2] += 1
	}

	s1 := 0
	s2 := 0
	for n, i := range nums1 {
		s1 += abs(i - nums2[n])
		if i < minNum2 {
			continue
		}
		s2 += i * counts[i-minNum2]
	}

	fmt.Println("Part 1:", s1)
	fmt.Println("Part 2:", s2)
}

func abs(i int) int {
	if i < 0 {
		return -i
	}
	return i
}

File: day01/src/main.rs
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut nums1 = Vec::new();
    let mut nums2 = Vec::new();
    for line in reader.lines() {
        let line = line?;
        let mut fields = line.split_whitespace();
        let num1 = fields.next().ok_or("Invalid")?.parse::<usize>()?;
        let num2 = fields.next().ok_or("Invalid")?.parse::<usize>()?;
        nums1.push(num1);
        nums2.push(num2);
    }
    nums1.sort();
    nums2.sort();

    let min_num2 = nums2[0];
    let max_num2 = nums2[nums2.len() - 1];

    let mut counts = vec![0; max_num2 - min_num2 + 1];
    for i in nums2.iter() {
        counts[i - min_num2] += 1;
    }

    let (s1, s2) = nums1
        .iter()
        .zip(nums2.iter())
        .fold((0, 0), |(s1, s2), (&v1, &v2)| {
            (
                s1 + v1.abs_diff(v2),
                s2 + if v1 < min_num2 {
                    0
                } else {
                    v1 * counts[v1 - min_num2]
                },
            )
        });

    println!("Part 1: {}", s1);
    println!("Part 2: {}", s2);
    Ok(())
}
